### PR TITLE
Refactor Profile Page props to work in three possible modes

### LINKED ISSUE #16: Refactor the profilePage prop
# Summary
The current profile page is not migrated properly to the new version of react-admin. We need to refactor it and also change the implementation to feature easier customisation. 

# Tasks
- [ ] Make the `profilePage` props work in three possible modes
   - [x] default value should be `true` - in such case we should use the default `<ProfilePage />` component
   - [x] `false` boolean value - this indicated that no profile page is going to be shown and we should omit the entry in the `userMenu`
   - [ ] it should support a custom component passed - this should still add the entry in the `userMenu` but the profile page route should load this component instead of the default one
 - [ ] Add a link to the profile page route in the `userMenu` in case the prop has value different from `false`
 - [ ] Include a custom `noLayout` route inside the `CustomRoutes` in `<AuthAdmin />` component for the profile page in case it is enabled
 
 # Extract page's logic to authProvider verbs
 - [ ] Add new verbs for each action on the profile page
 - [ ] Inside the profile page call all business logic through the `authProvider` 
 - [ ] Add example in the demo project
 - [ ] Document all the above changes

### PR DESCRIPTION
**Pull request by batanoffs**
### Summary
Changes made according to this [task](https://github.com/ValentinnDimitroff/ra-auth-ui/issues/16). Also added `defaultPasswordRules` (not fully implemented yet, future update in next pull request).
- `profilePage` now supports three possible modes: `default`, `boolean`, and `component`. 
- Route has been added for the custom profile component
- Demo added in examples 

### Detailed Description

**IMPORTANT** Please consider the import for `AuthAdmin` needs to be changed for production. Comment has been added.

The following changes have been made:

- added `profilePage` item into `userMenu` dropdown, complemented with conditionals
- update types and defaults for `authOptions` with new `profilePage` and `passwordRules` (password rules not implemented yet)
- added default values for `defaultPasswordRules`
- update `AuthOptionsContextType` to support `profilePage` options
- added route for profilePage 
```jsx
    <Route
          path={PROFILE_ROUTE}
          element={
              authOptions.profilePage === true ? (
                  <ProfilePage />
              ) : authOptions.profilePage === false ? null : (
                  authOptions.profilePage
              )
          }
      />
```
- update `defaultAuthOptions` to support `profilePage` props and fix children's type error for typescript
- add `TestProfile.tsx` in examples to test custom components.

### Checklist
- I have performed a self-review of my code, double checked and tested with an example demo.
- My changes generate no new warnings.
- Commented parts of my code for more clarity.

### CODE REVIEW

**Review by ValentinnDimitroff — CHANGES_REQUESTED** [review]
Overall looks pretty good - some minor comments to polish it and also one note: With each change/addition of new feature, we need to make sure the Readme is up to date so probably a good idea to add the following changes as a description there in the current PR.

File: src/pages/profile/ProfileDetailsCard.tsx [hunk 1]
```diff
@@ -16,6 +16,15 @@ const styles = {
     },
 }
 
+export type PasswordRulesType = {
+    RequiteDigit?: boolean
```
[the remark this answers is not in the document]
Comment (batanoffs) [reply src/pages/profile/ProfileDetailsCard.tsx:20]: I will check this

**Comment by batanoffs** [comment]
Updated according to the requested changes. Please review the new changes. I have also resolved some conflicts for the readme file.

File: src/utils.ts [hunk 2]
```diff
@@ -24,3 +19,21 @@ export const defaultAuthRoutes: { path: string; Component: React.FC }[] = [
...
     },
 ]
+
+export type PasswordRulesType = {
+    requiredDigit?: boolean
+    requiredLowercase?: boolean
+    requiredNonAlphanumeric?: boolean
+    requiredUppercase?: boolean
+    requiredLength?: number
+    requiredSymbols?: boolean
```
Comment (ValentinnDimitroff) [inline src/utils.ts]: What's the difference between `requiredSymbols` and `requiredNonAlphanumeric` ?

File: src/AuthAdmin.tsx [hunk 3]
```diff
@@ -27,16 +33,16 @@ export const AuthAdmin: FC<AuthAdminProps> = ({
...
+    // Note that we need to provide defaults for ProfilePage props onPictureUpload, onSubmit
+    const ProfileComponent = authOptions.profilePage === true ? <ProfilePage /> : authOptions.profilePage
     return (
         <AuthOptionsContextProvider {...authOptions}>
             <Admin loginPage={false} layout={layout} {...rest}>
                 <CustomRoutes noLayout>
                     {authRoutes.map(({ path, Component }) => (
                         <Route key={path} path={path} element={<Component />} />
                     ))}
+                    {authOptions.profilePage !== false && <Route path={PROFILE_ROUTE} element={ProfileComponent} />}
```
Comment (ValentinnDimitroff) [inline src/AuthAdmin.tsx]: If someone passes a `null` value for the `profilePage` prop - I believe this code will produce a route which leads to a null component

Comment (ValentinnDimitroff) [inline src/AuthAdmin.tsx]:
```suggestion
                    {authOptions.profilePage && <Route path={PROFILE_ROUTE} element={ProfileComponent} />}
```

File: src/utils.ts (hunk 2 above)
Comment (batanoffs) [reply src/utils.ts]: My idea was to group them like punctuation characters ! @ # & ( ) – [ { } ] : ; ', ? / * and symbol characters \` ~ $ ^ + = < > “. But I see the confusion here. I forgot to update the naming properly and so on. On second review, it's probably better to use only one rule, for example `requiredSymbols`. it's easier to understand.

File: src/AuthAdmin.tsx (hunk 3 above)
Comment (batanoffs) [reply src/AuthAdmin.tsx]: Yes, updated

File: README.md [hunk 4]
```diff
@@ -57,18 +56,21 @@ yarn install
...
 ```
 
-# Available Props
+## Available Props
+
+
+| Name        | Type              | Default   | Description                                                                                             |
+| ----------- | ----------------- | --------- | ------------------------------------------------------------------------------------------------------- |
+| `authRoutes`  | array             | array     | *If unset provides default authentication routes and screens. But you can pass custom routes like `{ path: string; Component: FC }`*                                          |
+| `authLayout`  | object            | undefined | *If set to '{userMenu: true}' provides default UserMenu. If set to object pass on the props to Layout.* |
```
Comment (ValentinnDimitroff) [inline README.md]: The prop is called `authOptions`.  It should point to the details of how the prop works, we can't explain all the features in the box here.

```suggestion
| `authOptions`  | object            | undefined | *If unset provides default Profile Page, integrated User Menu and password rules out of the box. All supported fields explained [below](#functionality-of-authlayout)* |
```

File: README.md [hunk 5]
```diff
@@ -118,72 +120,80 @@ const App = () => {
 
 ```
 Example
-``` 
-
-Docs Sketches ->
+``` -->
 
-## authLayout
+## Functionality of authLayout
```
Comment (ValentinnDimitroff) [inline README.md]:
```suggestion
## Functionality of authOptions
```

File: README.md [hunk 6]
```diff
@@ -118,72 +120,80 @@ const App = () => {
...
+ Provides authenticated layout with smart defaults.
 
-Provides authenticated layout with smart defaults.
+- ### Defaults 
+    The default props can be set by `authOptions`
 
-### Default
+    | Props         | Type    | Defaults    | Behavior                                                                                                                           |
+    | ------------- | ------- | ----------- | ---------------------------------------------------------------------------------------------------------------------------------- |
+    | profilePage   | boolean | true        | *By default profile page is enabled and a basic profile component is set.*                                                         |
```
Comment (ValentinnDimitroff) [inline README.md]:
```suggestion
    | profilePage   | boolean | Component | true        | *By default profile page is enabled and a basic profile component is set. You can opt out by setting it to `false` or pass your own component which will be rendered at the `/profile` route*                                                         |
```

File: README.md [hunk 7]
```diff
@@ -118,72 +120,80 @@ const App = () => {
...
-## authLayout
+## Functionality of authLayout
+ Provides authenticated layout with smart defaults.
 
-Provides authenticated layout with smart defaults.
+- ### Defaults 
+    The default props can be set by `authOptions`
 
-### Default
+    | Props         | Type    | Defaults    | Behavior                                                                                                                           |
```
Comment (ValentinnDimitroff) [inline README.md]: As a rule of thumb we can copy the react-admin style of describing props: https://marmelab.com/react-admin/Admin.html#props

```suggestion
    | Props         | Type    | Defaults    | Description                                                                                                                           |
```

File: README.md [hunk 8]
```diff
@@ -118,72 +120,80 @@ const App = () => {
...
+    | ------------- | ------- | ----------- | ---------------------------------------------------------------------------------------------------------------------------------- |
+    | profilePage   | boolean | true        | *By default profile page is enabled and a basic profile component is set.*                                                         |
+    | userMenuItems | array   | empty array | *Not specified.*                                                                                                                   |
```
Comment (ValentinnDimitroff) [inline README.md]:
```suggestion
    | userMenuItems | array   | empty array | *Populates the user menu with items depending on the `authOptions` configuration*                                                                                                                   |
```

File: README.md [hunk 9]
```diff
@@ -118,72 +120,80 @@ const App = () => {
...
+  *Note: The default routes can be changed*
+  
+- ### authOptions
 
-### profilePage
+  If unset provides default user menu ready to use. But you can pass custom profilePage, loginRedirectPath or array of userMenuItems to the user menu.
 
--   default is provided
--   supports passing own component
+- ### profilePage
```
Comment (ValentinnDimitroff) [inline README.md]: This is duplicated and outdated

File: README.md [hunk 10]
```diff
@@ -118,72 +120,80 @@ const App = () => {
...
+  -   userMenu - if no `appBar` is passed
+      -   true - default menu with profile link and logout button
+      -   array of object - {to, ...}
+      -   array of elements - TODO fnc receiving permissions and return element
 
-Other supported keys
+  - Other supported keys
+      -   menu
+      -   appBar
+      -   sideBar
```
Comment (ValentinnDimitroff) [inline README.md]: This is outdated

File: README.md [hunk 11]
```diff
@@ -118,72 +120,80 @@ const App = () => {
...
-Provides authenticated layout with smart defaults.
+- ### Defaults 
+    The default props can be set by `authOptions`
 
-### Default
+    | Props         | Type    | Defaults    | Behavior                                                                                                                           |
+    | ------------- | ------- | ----------- | ---------------------------------------------------------------------------------------------------------------------------------- |
+    | profilePage   | boolean | true        | *By default profile page is enabled and a basic profile component is set.*                                                         |
+    | userMenuItems | array   | empty array | *Not specified.*                                                                                                                   |
+    | passwordRules | object  | all enabled | *By default all rules are enabled. A digit, lowercase, non-alphanumeric,uppercase,min-length is `8` and min one symbol is needed.* |
```
Comment (ValentinnDimitroff) [inline README.md]: `loginRedirectPath` is missing in the table

**Review by ValentinnDimitroff — COMMENTED** [review]
Only readme related comments

File: src/utils.ts [hunk 12]
```diff
@@ -24,3 +19,19 @@ export const defaultAuthRoutes: { path: string; Component: React.FC }[] = [
         Component: ResetPasswordPage,
     },
 ]
+
+export type PasswordRulesType = {
+    requiredDigit?: boolean
+    requiredLowercase?: boolean
+    requiredUppercase?: boolean
+    requiredLength?: number
+    requiredMinSymbols?: number
```
Comment (ValentinnDimitroff) [inline src/utils.ts]: I actually liked the `requireNonAlphanumeric` prop better. Also having it as a bool property is enough e.g. having any symbol or none

File: src/AuthAdmin.tsx [hunk 13]
```diff
@@ -1,20 +1,26 @@
...
     authRoutes?: { path: string; Component: FC }[]
     authOptions?: AuthOptionsContextType
+    children?: ReactNode //fix typescript issues
 }
 
-export const defaultAuthOptions = { profilePage: <ProfilePage /> || null, userMenuItems: [] }
+export const defaultAuthOptions = {
+    profilePage: true,
+    userMenuItems: [],
+    passwordRules: defaultPasswordRules || null, // TODO - not yet implemented
```
Comment (ValentinnDimitroff) [inline src/AuthAdmin.tsx]: There is no need of `||` - `defaultPasswordRules` is a constant expression, it's never going to be null

File: prettier.config.js [hunk 14]
```diff
@@ -3,7 +3,7 @@ module.exports = {
     tabWidth: 4,
     semi: false,
     singleQuote: true,
-    printWidth: 100,
+    printWidth: 150,
```
Comment (ValentinnDimitroff) [inline prettier.config.js]: Why?

File: README.md (hunk 4 above)
Comment (batanoffs) [reply README.md]: Fair, will update accordingly.

File: src/utils.ts (hunk 12 above)
Comment (batanoffs) [reply src/utils.ts]: okey, will update it

File: prettier.config.js (hunk 14 above)
Comment (batanoffs) [reply prettier.config.js]: Helps me read better, but I can work with 100 if you prefer. It's not a problem

File: README.md (hunk 10 above)
Comment (batanoffs) [reply README.md]: removed

**Review by ValentinnDimitroff — APPROVED** [review]

**Pull request merged**